Sources/AWSSDKSwiftCore/MetaDataService.swift

```diff
@@ -180,32 +184,48 @@ struct InstanceMetaDataServiceProvider: MetaDataServiceProvider {
 
     typealias MetaData = InstanceMetaData
 
+    static let instanceMetadataApiTokenUri = "/latest/api/token"
     static let instanceMetadataUri = "/latest/meta-data/iam/security-credentials/"
     static var host = "169.254.169.254"
+    static var apiTokenURL: String {
+        return "http://\(host)\(instanceMetadataApiTokenUri)"
+    }
     static var baseURLString: String {
         return "http://\(host)\(instanceMetadataUri)"
     }
 
-    func uri(eventLoopGroup: EventLoopGroup) -> EventLoopFuture<String> {
+    func getCredential(eventLoopGroup: EventLoopGroup) -> EventLoopFuture<CredentialProvider> {
+        //  no point storing the session key as the credentials last as long
+        var sessionTokenHeader: [String: String] = [:]
         // instance service expects absoluteString as uri...
-        return request(uri:InstanceMetaDataServiceProvider.baseURLString, timeout: 2, eventLoopGroup: eventLoopGroup)
-            .flatMapThrowing{ response in
-                switch response.head.status {
-                case .ok:
-                    let roleName = String(data: response.body, encoding: .utf8) ?? ""
-                    return "\(InstanceMetaDataServiceProvider.baseURLString)/\(roleName)"
-                default:
-                    throw MetaDataServiceError.couldNotGetInstanceRoleName
+        return request(uri:InstanceMetaDataServiceProvider.apiTokenURL, method: .PUT, headers:["X-aws-ec2-metadata-token-ttl-seconds":"21600"], timeout: 2, eventLoopGroup: eventLoopGroup).flatMapThrowing { response in
+                // extract session key from response.
+                if response.head.status == .ok,
+                    let token = String(data: response.body, encoding: .utf8) {
+                    sessionTokenHeader = ["X-aws-ec2-metadata-token":token]
+                } else {
+                    print("Status \(response.head.status)")
                 }
-        }
-    }
-
-    func getCredential(eventLoopGroup: EventLoopGroup) -> EventLoopFuture<CredentialProvider> {
-        return uri(eventLoopGroup: eventLoopGroup)
-            .flatMap { uri in
-                return self.request(uri: uri, timeout: 2, eventLoopGroup: eventLoopGroup)
-            }
-            .map { response in
+                return
+            }.flatMapError { error in
+                // If we didn't find a session key then assume we are running IMDSv1 (we could be running from a Docker container
+                // and the hop count for the PUT request is still set to 1)
+                print(error)
+                return eventLoopGroup.next().makeSucceededFuture(())
```

Comment:
flatMapError to optional headers -> means we don't have headers here -> return nil